### PR TITLE
Testflo update

### PR DESCRIPTION
**Pull request by joanibal**
## Purpose
The goal the this PR is get further feedback (hence the draft) and discuss future development

All of the old test have been converted to a format usable by testflo.
BUT, there is a bug with testflo for test with the same name in different folders. I've fix this issue for my local testflo, but i need to upgrade to the latest version of testflo (python 3 only) to see if the problem still persist before making a PR.
Additionally, all the viscous functionals test do not pass because of issue #44.
Regardless there are some points I'd like to discuss before moving on.

1) The structure of the tests (please let me know what you think)

Ideas to improve tests
2) Change the solution tests to check that the residual are below a threshold not that they match the previous value.
3) update the solver options to speed up solution tests.
4) Change the time accurate case to euler to save run time. 
5) use frame work to add more adjoint and solution tests to existing regression tests
6) add more regression tests for...
	- low speed case (with lowspeed pc)
	- hydrofoil 

## Type of change
- New feature (non-breaking change which adds functionality)

## Checklist
_Put an `x` in the boxes that apply._

- [x] I have added new tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by nwu63** [comment]
Did you add new stuff to `BaseRegTest`? Would you mind pushing what you have to a branch under `mdolab`? I'd like to work with you to set a good API for that and have everything in place. That way others can start working on tests in other repos.

Also, what did you think of the idea of having a dedicated `testing` module? I would prefer to organize the stuff in `baseclasses` a bit more.

**Comment by joanibal** [comment]
I did add some stuff to baseclasses. I changed it to allow comparison of arrays directly instead of comparing each element. I also added static methods for reading and writting jsons.

After I clean it up I'll make a pull request on the mdolab repo

File: tests/reg_tests/convert_old_refs.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import glob
```
[the remark this answers is not in the document]
Comment (joanibal) [reply tests/reg_tests/convert_old_refs.py:1]: I used the function to generate the new reference files. So it does work. It can break when things are printed out of order because of some MPI shenanigans, but that only happened once in the 18 adflow reg test.

This function should probably go in the centralized registration testing modules.

File: tests/reg_tests/euler/scalar_jst/test_CRM_WBT.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+# import the ADFLOW module this test lives in 
+from python.pyADflow import ADFLOW
+
+# import the testing utilities that live a few directories up 
+import reg_test_utils as utils
+
+from reg_default_options import adflowDefOpts, defaultAeroDVs
+
+from reg_aeroproblems import ap_CRM 
+from reg_test_classes import SolveRegTest, FunctionalsRegTest
```
[the remark this answers is not in the document]
Comment (joanibal) [reply tests/reg_tests/euler/scalar_jst/test_CRM_WBT.py:25]: Each of those classes inherit from `RegTest`,  which could be reused for IDwarp.
https://github.com/joanibal/adflow/blob/b6e472a85caeb92ca32bfc599be150805edc05c3/tests/reg_tests/reg_test_classes.py#L17

**Comment by joanibal** [comment]
@nwu63  and @anilyil 
What do you think about ideas 2-5?

**Comment by nwu63** [comment]
I'm not very familiar with ADflow so these are my uneducated comments:

2. Not sure what you mean. Do we not currently converge solution fully, then compare outputs like cl/cd?
3. Agreed, as long as it does not reduce our test coverage.
4. I guess this is okay? The time stepping is decoupled from the spatial solution right?
5. Definitely
6. What about rotating cases? Supersonic flow? Checking output quantities such as separation sensor? I'm sure there are lots that we are not testing.

What about unit tests? Is enough stuff exposed to the python layer to allow us to do that? We should also figure out other combinations of options to test, in order to expand the test coverage.

**Comment by anilyil** [comment]
My feedback for the listed items (and for some additional stuff):

1. I will take a look once I can run the code.
2. @nwu63 I think he meant that prescribing a 1e-12 convergence for example, and comparing the residual norm with the reference, and not comparing the full residual vector. @joanibal I think this makes sense for some of the tests, but I would rather keep one test with a full residual comparison after convergence, just as a baseline. All of the cases that I can think of where the residual norm test would pass but a direct residual test would fair are very weird edge cases, so probably not worth checking for. Comparing just the residual norm has the advantage of speeding up future solver development, as things will converge from a different direction with any changes to the solver.
3. I am fine with this, as long as you clearly state the previous and current state. I think only the unsteady test takes relatively long, the others are probably fine. As a different perspective, I would rather keep the current setup just to test a wide range of solver options. But then again, we can create new tests that cover these combinations.
4. Does the turbulence or the viscous fluxes get any special treatment for an unsteady case? I think at least turbulence does, and possibly the viscous fluxes. I say we keep unsteady RANS to track that part of the code.
5. I am all for additional solver and adjoint tests. We can test a range of things, and possibly add tests that mimic the MPhys wrapper, so we also keep a functional wrapper at all times. Current om_adflow wrapper only gets fixes when someone needs to use it.
6. I dont think lowspeed PC is functional. Then again, I am not sure if @yqliaohk is using it for her cases. If she is using the lowspeedPC, then we can add tests, if not, possibly not worth it. I am definitely for adding at least one low-speed case. At least to document the flow conditions. Maybe we can create a very small mesh to test this.

Additional tests I recommend adding: I think we should have a range of tests for blockettes. Because we are planning to make blockettes disabled by default, we need to explicitly test the blockette residual routines. For these, I think we can compare the block residuals to blockette residuals, and test various discretization combinations.

Besides this, should we run and test these ourselves? Is the code ready? Are there any other relevant PRs open on other repos that we need to run these?

**Comment by joanibal** [comment]
2) We have never compared the full residuals vectors in any regression test, It has always been norms. In the old reg tests we compare the norm of the residual against the prior value. I'm suggesting that we change it to compare the norm of the residuals against a threshold. This way, the solve test will not fail if the converge options change. 

The code can't be run now because of difficulties of testflo **with the structure of the test**.
Specifically having test files with the same name in different folders. 
So if you don't like the structure, i won't bother fixing the testflo issue (though it should probably be fixed anyways).

**Comment by anilyil** [comment]
Okay, makes sense for item 2. I misunderstood.

I am fine with the structure. I think the testflo issue should be fixed anyways.

**Comment by nwu63** [comment]
I guess I'm a little confused about the test structure. For example in `tests/reg_tests/euler/matrix_jst/test_tutorial_wing.py`, where is the actual call for the test? There is only a `setUp` function which does not do any comparisons or anything as far as I could tell.

My other question was, do we have enough tests that are largely the same, but with slightly different options, that we could take advantage of the `parametrized` package to do parameterized testing? I think that would be really good in reducing code duplication. If that is done, then the location of the tests and the overall directory structure will have to be changed.

**Comment by joanibal** [comment]
The actual test_blah methods for each class are inherited from `SolveTest`, `FunctionalsTest`, or `AdjointTest`. 

I'd be interested in combined tests using `parameterize`. Do you have an example of that working with testflo?

**Comment by nwu63** [comment]
I see, so if you inherit from a class that implements a "test_" method, then that will automatically get picked up by testflo? I guess that makes sense, thanks.

There was some back/forth discussion on the testflo repo, where they just said they support using the package `parametrized`, so I assume you would use it normally? Basically you specify parameters via decorators, and that would invoke multiple calls to the function. Their [docs](https://github.com/wolever/parameterized) are pretty good.

Hopefully in the future they also add support for providing a list to the `N_PROCS` argument, which can be used to parameterize both serial and parallel tests without code duplication.

**Comment by joanibal** [comment]
@nwu63 
it' ready for review
There are a few small hacks to get `parameterized` to work for our uses, but it makes other things much simpler.

**Comment by joanibal** [comment]
Tests are failing because we need a new `Baseclasses` on our docker image

**Comment by nwu63** [comment]
I will finish up that PR hopefully tonight, then these tests should pass once the docker images are updated.

In the mean time, could you summarize what you did in terms of test coverage? Are these the same tests that ran previously, or have you added anything new?

I will take a closer look at the PR and provide detailed feedback soon.

**Comment by nwu63** [comment]
Travis sees failed tests as passing because the Python testing script uses an `os.system` call, which does not forward the exit status. I think we want to either modify the Python script to do that, or just hard code the test command in Travis.

I am slightly in favour of the second option, and instead just run `testflo` in the root directory. That way it's clear what is being done, and I don't anticipate the testing command changing much. Was there a reason to use `-n 2` by the way?

**Comment by joanibal** [comment]
Sorry it took a week, but here are my answers

>In the mean time, could you summarize what you did in terms of test coverage? Are these the same tests that ran previously, or have you added anything new?

The test coverage is almost the same. More tests can be added later, but it felt like it would be best to have another PR for those. The one exception is the cs tests, which include a few more functions, because functions are free in fwd mode and it was easier to write. 

>  I think we want to either modify the Python script to do that, or just hard code the test command in Travis.

The later sounds good to me too. 

>Was there a reason to use -n 2 by the way?

For some reason i thought the machines used by Travis had 2 cores

**Comment by nwu63** [comment]
Thanks for that @joanibal. I'll push some updates to travis/run script.

For some reason I was under the impression that `-n` was for number of tests to run in parallel rather than the number of procs. Does it override the values set in each test? What happens if a test needs 4 procs but you run it with `-n 2`?

**Comment by joanibal** [comment]
I did some quick tests and I think that you are right, `-n` doesn't appear to change the number of processors used for a tests.

When I run a test that uses 4 procs with `-n=2` it uses 4 procs.

**Comment by nwu63** [comment]
So a few tests failed for various reasons
- build 1 timed out, not sure why but the intel build is taking significantly longer, it got through most of the tests though
- builds 3/4 failed because the inheritance stuff (`.super().setUp()` syntax) is py3 only. Given we are removing py2 support, this is fine
- builds 7/8 failed because that image does not contain the updated baseclasses, due to test failure last night on the build bot. This can also be ignored

**Comment by joanibal** [comment]
@nwu63  Do you think we could increase the number of concurrent tests to make sure it doesn't time out?

**Comment by nwu63** [comment]
Are you talking about splitting it up into even more parallel jobs? So instead of running the real tests there would be two jobs each running like half the tests? It's possible, but we'd have to think about how to do this intelligently with testflo so that we don't miss anything. I'm curious why it's taking so long though, maybe we should do some profiling?

**Comment by joanibal** [comment]
I was thinking of increasing `-n` when calling testflo, but i see now that the default(the max) is already being used.

You bring up a much better point. We should probably figure out what is so slow. Maybe it's related to issue #67 ?

**Comment by nwu63** [comment]
I just cross-checked the Travis logs between the intel build and one of the ubuntu-gcc builds:
- even things that shouldn't matter, like downloading the input files, took twice as long on the intel real tests compared to the gcc real tests
- but for the same pair of images, the same operation for the complex tests were comparable

So based on that, maybe that was just one "bad" VM or something, even though the hardware looked identical -- both running on GCE. I will restart the test that timed out and see if it ran faster this time.

**Comment by nwu63** [comment]
Okay, I've talked to Travis support and increased the build time limit from 2 to 3 hours, but the Intel image is still timing out -- now at 79/80 tests. So close.. But the conclusion is that this is indeed taking much longer than the gcc-based images. I'm not sure if it's the compiler, or the OS, or some other reason, but IMO double the time is a bit ridiculous.

That being said, I am okay with disabling this test for now, and merging this in, as long as we open an issue and document this problem.

**Comment by joanibal** [comment]
This is so weird. Are unrelated tasks, like downloading files, still taking longer on the Intel image?

Did the Intel tests take longer when using the old regression tests framework?

**Comment by nwu63** [comment]
@eirikurj and I discussed this, and we're really not sure. The download time seems comparable now, but the run time is very long. We suspect it has to do with the intel compiler -- we are going to set up the centos7-gcc image tonight and have that available as a test, to try and narrow this down. We may also want to play around with the compiler flags used in these images.

**Comment by joanibal** [comment]
>we must be doing something wrong like virtualizing 4 procs or something.
Should I change all of the tests to run on 2 Procs?
    N_PROCS = 4 -> N_PROCS = 2

**Comment by nwu63** [comment]
Yeah let's try that. I just checked that testflo will try to request whatever you specify in `N_PROCS`, so it will oversubscribe. Maybe impi and openmpi handles this differently or something.

**Comment by nwu63** [comment]
@joanibal the new Docker image with the context manager will be available tomorrow, so I'll trigger a build and hopefully it won't time out anymore.

**Comment by nwu63** [comment]
So that one test fails with `rtol=1.2e-10` instead of `1e-10`. What do we want to do? Loosening the tolerance means we have to loosen the tolerance for all the `jac_vec_prod_fwd` tests, but maybe `5e-9` isn't too bad?

**Review by eirikurj — APPROVED** [review]
